test/jdk/com/sun/jdi/CLETest.java

```diff
@@ -0,0 +1,441 @@
+            stepRequest.enable();
+        }
+        // Complete test5. We fail if we never saw the expected co-located events.
+        if (breakpointCount == 8) {
+            if (testcaseFailed) {
+                testFailed = true;
+                System.out.println("TESTCASE #5 FAILED: did not get Step and Breakpoint co-located events");
+            } else {
+                System.out.println("TESTCASE #5 PASSED");
+            }
+        }
+
+        // Setup test: MethodEntry and Step co-located events
+        // Completion is handled by the stepCompleted() since there is no additional breakpoint.
+        if (breakpointCount == 9) {
+            testcaseFailed = true; // assume failing unless pass detected
+            entryRequest.enable();
+            stepRequest = erm.createStepRequest(mainThread,
+                                                StepRequest.STEP_LINE,
+                                                StepRequest.STEP_INTO);
+            stepRequest.addCountFilter(1);
+            stepRequest.setSuspendPolicy(EventRequest.SUSPEND_ALL);
+            stepRequest.enable();
+        }
+    }
+
+    public void stepCompleted(StepEvent event) {
+        System.out.println("Got StepEvent: " + getLocationString(event.location()));
+        event.request().disable();
+        entryRequest.disable();
+        if (testcase == 6 && testcaseFailed) {
+            testFailed = true;
+            System.out.println("TESTCASE #6 FAILED: did not get MethodEntry and Step co-located events");
+        }
+        if (testcase == 1 || testcase == 2 || testcase == 6) {
+            exitRequest.disable();
+            if (!testcaseFailed) {  // We already did a println if the test failed.
+                System.out.println("TESTCASE #" + testcase + " PASSED");
+            }
+        }
+    }
+
+    public void methodEntered(MethodEntryEvent event) {
+        System.out.println("Got MethodEntryEvent: " + getLocationString(event.location()));
+        if (methodEntryCount++ == 25) {
+            entryRequest.disable(); // Just in case the test loses control.
+        }
+    }
+
+    public void methodExited(MethodExitEvent event) {
+        System.out.println("Got MethodExitEvent: " + getLocationString(event.location()));
+        //printStack(event.thread());
+        exitRequest.disable();
+        entryRequest.disable();
+    }
+
+    protected void runTests() throws Exception {
+        System.out.println("Starting CLETest");
+        BreakpointEvent bpe = startToMain("CLEDebugee");
+        targetClass = (ClassType)bpe.location().declaringType();
+        mainThread = bpe.thread();
+        System.out.println("Got main thread: " + mainThread);
+        erm = eventRequestManager();
+
+        try {
+            // Setup all breakpoints
+            for (MethodBreakpointData bpData : breakpoints) {
+                Location loc = findMethodLocation(targetClass, bpData.method,
```

Comment:
Specify breakpoints by line number in a method is better than just line number, but did you think about mark them with some tag and parse source file to get line numbers? (like in test/jdk/com/sun/jdi/lib/jdb/JdbTest.java, parseBreakpoints method)

Reply:
It looks like the tests in com/sun/jdi are split between those that extend `JdbTest` and those that extend `TestScaffold`. I'm not too sure of the rationale for using one over the other, but I ended up with `TestScaffold` because of the test I initially used as a template (BreakpointTest.java). `parseBreakpoints()` is only supported by `JdbTest`, so I would need to convert the test to extend it instead. I have no idea how disruptive this might be to the test.

I think the breakpoints I have are fairly safe from code movement and modifications. It would require edits within the methods with the breakpoints, all of which are very simple and well commented, and there really is no reason to ever edit them in the first place. Also there are 8 other tests that are sensitive to line numbers (search for THIS TEST IS LINE NUMBER SENSITIVE), and they are at much greater risk (adding or removing a line anywhere before the breakpoint will break the test).

Reply:
AFAIR JdbTest is a base class to test jdb functionality, I didn't mean to extend it. But breakpoint parsing code is quite simple, I thought about add similar method to TestScaffold instead of  findMethodLocation.
As I wrote your way with findMethodLocation is much better than just line numbers (used in "THIS TEST IS LINE NUMBER SENSITIVE" tests), I'm okay with it.

Reply:
Ok. Thanks for the review!